perf improvements

rvagg opened the pull request:
Using [pino](https://github.com/mcollina/pino)'s own benchmark suite to add some perf improvements .. some are kind of crazy (there's a good reason I made a separate file called format.js). Mostly it's not down to the speed of pino, but it's at least competitive now.

API hasn't really changed, except now it's limited to 16 arguments on a call to a log function, I think this is more than reasonable but still it's technically a breaking change. There's also a new `bole.setFastTime()` function to switch to `Date.now()` which is the pino default (personally I hate logs that use epoch time but it is quite fast so we can give the option at least).

It's also very possible that there is accidental breakage in the API here given all of the shuffling of logic so bumping semver-major if we go ahead with this is probably wise regardless.

Below are benchmarks comparing old and new bole as well as bole with `fastTime` enabled. The benchmarks were run with: ```for i in `ls ../pino/benchmarks/`; do echo $i; node ../pino/benchmarks/$i ; done``` having checked pino out, done an `npm install` and linked bole into pino's node_modules. The `fastTime` ones were done by adding `.setFastTime(true)` to all of the bole setup sections in the files.

Happy to hear of more crazy ways we can speed this up! I'm not sold on pino's "extreme" mode fwiw, too much hairy logic to get that right.

/cc @yoshuawuyts 


```
basic.js
benchBunyan*10000:     1608.165ms
benchWinston*10000:    2582.654ms
benchOldBole*10000:    1774.625ms
benchNewBole*10000:    585.776ms
benchNewestBole*10000: 372.713ms
benchDebug*10000:      545.641ms
benchLogLevel*10000:   437.950ms
benchPino*10000:       370.170ms
benchPinoExreme*10000: 142.827ms
benchBunyan*10000:     1627.781ms
benchWinston*10000:    2638.942ms
benchOldBole*10000:    1637.006ms
benchNewBole*10000:    574.749ms
benchNewestBole*10000: 352.200ms
benchDebug*10000:      527.688ms
benchLogLevel*10000:   394.412ms
benchPino*10000:       324.713ms
benchPinoExreme*10000: 134.498ms


child.js
benchBunyanChild*10000:      1717.793ms
benchOldBoleChild*10000:     1625.616ms
benchNewBoleChild*10000:     616.279ms
benchNewestBoleChild*10000:  458.978ms
benchPinoChild*10000:        421.797ms
benchPinoExtremeChild*10000: 190.450ms
benchBunyanChild*10000:      1735.584ms
benchNewBoleChild*10000:     596.083ms
benchNewestBoleChild*10000:  439.364ms
benchPinoChild*10000:        405.750ms
benchPinoExtremeChild*10000: 174.887ms


childCreation.js
benchBunyanCreation*10000:      1912.047ms
benchNewBoleCreation*10000:     674.235ms
benchNewestBoleCreation*10000:  473.900ms
benchPinoCreation*10000:        455.381ms
benchPinoExtremeCreation*10000: 222.714ms
benchBunyanCreation*10000:      1958.595ms
benchOldBoleChild*10000:        1597.697ms
benchNewBoleCreation*10000:     636.196ms
benchNewestBoleCreation*10000:  443.747ms
benchPinoCreation*10000:        454.718ms
benchPinoExtremeCreation*10000: 204.472ms


deep-object.js
benchBunyanDeepObj*10000:            9922.489ms
benchWinstonDeepObj*10000:           20467.637ms
benchOldBoleDeepObj*10000:           22942.288ms
benchNewBoleDeepObj*10000:           5334.259ms
benchNewestBoleDeepObj*10000:        5976.224ms
benchLogLevelDeepObj*10000:          51442.827ms
benchPinoDeepObj*10000:              5829.747ms
benchPinoUnsafeDeepObj*10000:        4346.752ms
benchPinoExtremeDeepObj*10000:       5941.517ms
benchPinoUnsafeExtremeDeepObj*10000: 4320.007ms
benchBunyanDeepObj*10000:            10368.592ms
benchWinstonDeepObj*10000:           20480.133ms
benchOldBoleDeepObj*10000:           23617.181ms
benchNewBoleDeepObj*10000:           5171.666ms
benchNewestBoleDeepObj*10000:        5913.482ms
benchLogLevelDeepObj*10000:          51737.796ms
benchPinoDeepObj*10000:              5708.446ms
benchPinoUnsafeDeepObj*10000:        4365.840ms
benchPinoExtremeDeepObj*10000:       5984.172ms
benchPinoUnsafeExtremeDeepObj*10000: 4188.128ms


multiArg.js
benchBunyanMulti*10000:                       1515.135ms
benchWinstonMulti*10000:                      2647.043ms
benchOldBoleMulti*10000:                      1856.140ms
benchNewBoleMulti*10000:                      594.838ms
benchNewestBoleMulti*10000:                   377.063ms
benchLogLevelMulti*10000:                     444.528ms
benchDebugMulti*10000:                        516.222ms
benchPinoMulti*10000:                         360.566ms
benchPinoExtremeMulti*10000:                  150.123ms
benchDebugInterpolate*10000:                  665.639ms
benchBunyanInterpolate*10000:                 1533.101ms
benchWinstonInterpolate*10000:                2809.965ms
benchOldBoleInterpolate*10000:                1845.031ms
benchNewBoleInterpolate*10000:                588.014ms
benchNewestBoleInterpolate*10000:             366.402ms
benchPinoInterpolate*10000:                   491.378ms
benchPinoExtremeInterpolate*10000:            192.419ms
benchBunyanInterpolateAll*10000:              1780.023ms
benchDebugInterpolateAll*10000:               803.718ms
benchWinstonInterpolateAll*10000:             2956.432ms
benchOldBoleInterpolateAll*10000:             2055.106ms
benchNewBoleInterpolateAll*10000:             686.849ms
benchNewestBoleInterpolateAll*10000:          454.530ms
benchLogLevelInterpolateAll*10000:            623.155ms
benchPinoInterpolateAll*10000:                547.891ms
benchPinoExtremeInterpolateAll*10000:         303.352ms
benchPinoUnsafeInterpolateAll*10000:          565.622ms
benchPinoUnsafeExtremeInterpolateAll*10000:   309.048ms
benchDebugInterpolateExtra*10000:             807.428ms
benchBunyanInterpolateExtra*10000:            3179.967ms
benchWinstonInterpolateExtra*10000:           3169.942ms
benchOldBoleInterpolateExtra*10000:           3736.959ms
benchNewBoleInterpolateExtra*10000:           1926.395ms
benchNewestBoleInterpolateExtra*10000:        1548.289ms
benchLogLevelInterpolateExtra*10000:          1759.567ms
benchPinoInterpolateExtra*10000:              620.915ms
benchPinoUnsafeInterpolateExtra*10000:        628.565ms
benchPinoExtremeInterpolateExtra*10000:       397.313ms
benchPinoUnsafeExtremeInterpolateExtra*10000: 375.254ms
benchDebugInterpolateDeep*10000:              663.052ms
benchBunyanInterpolateDeep*10000:             6041.667ms
benchWinstonInterpolateDeep*10000:            22134.751ms
benchOldBoleInterpolateDeep*10000:            6587.924ms
benchNewBoleInterpolateDeep*10000:            4741.358ms
benchNewestBoleInterpolateDeep*10000:         4217.562ms
benchLogLevelInterpolateDeep*10000:           3323.441ms
benchPinoInterpolateDeep*10000:               6017.806ms
benchPinoUnsafeInterpolateDeep*10000:         4708.956ms
benchPinoExtremeInterpolateDeep*10000:        6533.090ms
benchPinoUnsafeExtremeInterpolateDeep*10000:  4726.833ms
benchBunyanMulti*10000:                       1556.186ms
benchWinstonMulti*10000:                      2641.214ms
benchOldBoleMulti*10000:                      1873.938ms
benchNewBoleMulti*10000:                      583.086ms
benchNewestBoleMulti*10000:                   355.907ms
benchLogLevelMulti*10000:                     470.015ms
benchDebugMulti*10000:                        518.127ms
benchPinoMulti*10000:                         358.284ms
benchPinoExtremeMulti*10000:                  150.430ms
benchDebugInterpolate*10000:                  666.039ms
benchBunyanInterpolate*10000:                 1544.669ms
benchWinstonInterpolate*10000:                2633.577ms
benchOldBoleInterpolate*10000:                1803.603ms
benchNewBoleInterpolate*10000:                579.716ms
benchNewestBoleInterpolate*10000:             368.567ms
benchPinoInterpolate*10000:                   400.438ms
benchPinoExtremeInterpolate*10000:            179.811ms
benchBunyanInterpolateAll*10000:              1734.928ms
benchDebugInterpolateAll*10000:               830.598ms
benchWinstonInterpolateAll*10000:             2914.816ms
benchOldBoleInterpolateAll*10000:             2108.744ms
benchNewBoleInterpolateAll*10000:             668.581ms
benchNewestBoleInterpolateAll*10000:          459.191ms
benchLogLevelInterpolateAll*10000:            564.736ms
benchPinoInterpolateAll*10000:                566.521ms
benchPinoExtremeInterpolateAll*10000:         314.738ms
benchPinoUnsafeInterpolateAll*10000:          552.270ms
benchPinoUnsafeExtremeInterpolateAll*10000:   299.188ms
benchDebugInterpolateExtra*10000:             805.610ms
benchBunyanInterpolateExtra*10000:            3229.008ms
benchWinstonInterpolateExtra*10000:           3169.397ms
benchOldBoleInterpolateExtra*10000:           3462.622ms
benchNewBoleInterpolateExtra*10000:           1865.117ms
benchNewestBoleInterpolateExtra*10000:        1549.994ms
benchLogLevelInterpolateExtra*10000:          1897.569ms
benchPinoInterpolateExtra*10000:              642.484ms
benchPinoUnsafeInterpolateExtra*10000:        623.055ms
benchPinoExtremeInterpolateExtra*10000:       394.064ms
benchPinoUnsafeExtremeInterpolateExtra*10000: 365.378ms
benchDebugInterpolateDeep*10000:              667.786ms
benchBunyanInterpolateDeep*10000:             6059.128ms
benchWinstonInterpolateDeep*10000:            22247.574ms
benchOldBoleInterpolateDeep*10000:            6609.189ms
benchNewBoleInterpolateDeep*10000:            4562.093ms
benchNewestBoleInterpolateDeep*10000:         4262.331ms
benchLogLevelInterpolateDeep*10000:           3293.833ms
benchPinoInterpolateDeep*10000:               6101.653ms
benchPinoUnsafeInterpolateDeep*10000:         4854.862ms
benchPinoExtremeInterpolateDeep*10000:        7011.338ms
benchPinoUnsafeExtremeInterpolateDeep*10000:  4695.801ms


object.js
benchBunyanObj*10000:                         1678.507ms
benchWinstonObj*10000:                        2656.642ms
benchOldBoleObj*10000:                        1588.992ms
benchNewBoleObj*10000:                        597.381ms
benchNewestBoleObj*10000:                     415.290ms
benchLogLevelObject*10000:                    1862.047ms
benchPinoObj*10000:                           398.471ms
benchPinoUnsafeObj*10000:                     461.343ms
benchPinoExtremeObj*10000:                    205.798ms
benchPinoUnsafeExtremeObj*10000:              191.235ms
benchBunyanObj*10000:                         1687.752ms
benchWinstonObj*10000:                        2791.879ms
benchOldBoleObj*10000:                        1562.442ms
benchNewBoleObj*10000:                        586.917ms
benchNewestBoleObj*10000:                     380.834ms
benchLogLevelObject*10000:                    1945.977ms
benchPinoObj*10000:                           368.611ms
benchPinoUnsafeObj*10000:                     476.259ms
benchPinoExtremeObj*10000:                    238.516ms
benchPinoUnsafeExtremeObj*10000:              200.529ms
```

rvagg commented:
fwiw I tried to do some manual JSON strinficiation, creating the initial part of the object and tacking on a stringification of the rest of the object if there is any, but it resulted in a slowdown, even for the simple case where you don't need to add any additional stringification to the manual bit. I also tried out pino's "faster" util.format but it didn't help at all.

I'm doing this all on Node 6, perhaps it's different on older versions.

Fishrock123 commented on format.js:
```diff
@@ -0,0 +1,58 @@
+var utilformat = require('util').format
+
+
+function format (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16) {
```
switch on `arguments.length`?

thlorenz replied:
querying `arguments.length` makes things slower afaik.

yoshuawuyts replied:
that's what I've found as well

Fishrock123 replied:
It *shouldn't* so long as you don't do re-assigns on named arguments...

We [do this in core](https://github.com/nodejs/node/blob/master/lib/timers.js#L383-L407) and I feel like you might make it back by using the switch since it might optimize better.

Also, keeping this under 50 LOC so that it is inline-able will definitely help.

Fishrock123 replied:
Also if you cache `arguments.length` rather than putting it in the switch, that may also help. Hard to know without looking at IRHydra.

rvagg replied:
yeah, I tinkered around a lot with `arguments`, the perf gain is real but small enough to make all of this potentially more hassle than it's worth. Ultimately I think this is a distraction from the bigger perf gains.

brycebaril commented on format.js:
```diff
@@ -0,0 +1,58 @@
+}
+
+
+module.exports = format
```
I feel like
```js
function format (a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16) {
  if (a16 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16)
  if (a15 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15)
  if (a14 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14)
  if (a13 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13)
  if (a12 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12)
  if (a11 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11)
  if (a10 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)
  if (a9 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8, a9)
  if (a8 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7, a8)
  if (a7 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6, a7)
  if (a6 !== undefined)
    return utilformat(a1, a2, a3, a4, a5, a6)
  if (a5 !== undefined)
    return utilformat(a1, a2, a3, a4, a5)
  if (a4 !== undefined)
    return utilformat(a1, a2, a3, a4)
  if (a3 !== undefined)
    return utilformat(a1, a2, a3)
  if (a2 !== undefined)
    return utilformat(a1, a2)
  if (a1 !== undefined)
    return utilformat(a1)
  return undefined
}
```
Is a much cleaner way to write this. Not sure there would be a perf difference, though.

rvagg replied:
aye, good call

brycebaril commented:
I made a bole-only version of one of the benchmarks: [multiArg.js](https://dl.dropboxusercontent.com/u/6789610/bole/multiArg.js)

Here's the [v8-profiler output](https://dl.dropboxusercontent.com/u/6789610/bole/bench.cpuprofile)

The majority of the time is spent in bole's `namedLevelLogger` after that it's mostly formatting.

Here's the IRHydra artifacts: [code.asm](https://dl.dropboxusercontent.com/u/6789610/bole/code.asm) and [hydrogen-30454-1.cfg](https://dl.dropboxusercontent.com/u/6789610/bole/hydrogen-30454-1.cfg)

IRHydra doesn't show any deopts in `namedLevelLogger` or any other part of `bole` but it does show lots of deopts in core. (timers, buffer, and streams)

There were some deopts in `util.format` and associated code, so bypassing that entirely could be one approach, though based on the v8-profiler output, something to remove instruction count in `namedLevelLogger` would be where I'd start.

brycebaril commented:
I also generated a flamegraph you can find here: http://brycebaril.com/profile-32255/flamegraph.html

Definitely suggests looking at ways to replace or work better with `util.format` but there are some other potential Buffer integrations to consider. Also JSON.stringify is a hit.

yoshuawuyts commented:
`pino` uses [quick-format](https://www.npmjs.com/package/quick-format), it might perhaps work as a proper substitution for `util.format` here

rvagg commented:
I did originally try quick-format but it _added_ time rather than making it any quicker, fast-safe-stringify is in this PR and fast-json-parse isn't necessary for bole.

rvagg commented:
OK folks, I've _mostly_ cracked it. It seems that the bulk of the time is spent in `stringify()`, I was thinking that a lot of the cost was in `write()`, so I revisited the partial manual stringification and did it a tiny bit differently than before, plus I added a bunch of other micro optimisations in here and I've updated the benchmarks in the OP with new numbers for `boleNewest` (also updated the Pino numbers so they are from the same run. You can see that we are within sneezing distance for the basic benchmark. I don't know where else I can shave off time and am not really sure what the edge is that pino has left here.

Interestingly, if you look at the interpolation tests, we're doing very well, and in some we're beating pino by a non-trivial amount. I think the cost for pino here is in `quick-format` because if I switch out at this stage, bole slows down by about that much.

The object benchmarks show a roughly similar difference between bole and pino as the basic ones, since objectMode bypasses the stringification completely, I believe this says that I've managed to completely remove the difference between them in stringification (except for deeper object I suspect). So there's something left here!

If nothing else comes of this I'm going to publish as a semver-major and submit a PR to pino with an update to the version & bole setup in their benchmarks for fastTime and suggest they change the language to "pino is _one of_ the fastest logging libraries", just for fun!

rvagg commented:
weird, I believe I've just reduced the difference in the basic.js benchmark to be statistically insignificant _with this one weird trick_:

```
diff --git a/bole.js b/bole.js
index 015d5f8..d76ae67 100644
--- a/bole.js
+++ b/bole.js
@@ -88,7 +88,11 @@ function stringify (level, name, message, obj) {
   for (k in obj)
     s += ',' + _stringify(k) + ':' + _stringify(obj[k])

-  return s + '}'
+  s += '}'
+
+  Number(s) // convert internal representation to plain string
+
+  return s
 }
```

Thanks to the hint from https://github.com/davidmarkclements/flatstr which pino uses. Have a read of the README there, I didn't know this about concatenated strings.

rvagg commented:
then .. applying the same trick to a pre-cached section of the stringified output (hostname, pid, level) and using that instead of the extra concatenation when performing stringification .. gets us to a statistically significant winning position in the basic! pino would have difficulty with this because it allows you to change the hostname and its output is jumbled up with the potentially cachable bits not next to eachother so it'd be a breaking change (sort of) if they were to do it with the tiny bit that they could do it with.

Fishrock123 commented:
Wow, that is very interesting!

rvagg commented:
merged and released as v3, but the challenge still exists if anyone wants to try and squeeze more out of this

yoshuawuyts commented:
Maddd!
